rewrite-java-17/src/main/java/org/openrewrite/java/isolated/ReloadableJava17TypeMapping.java

```diff
@@ -692,22 +691,65 @@ private void completeClassSymbol(Symbol.ClassSymbol classSymbol) {
-                if (retention != null && retention.value() == RetentionPolicy.SOURCE) {
-                    continue;
-                }
-                annotations.add(annotType);
+        if (!sym.getDeclarationAttributes().isEmpty()) {
+            annotations = new ArrayList<>(sym.getDeclarationAttributes().size());
+            for (Attribute.Compound a : sym.getDeclarationAttributes()) {
+                JavaType.Annotation annotation = annotationValue(a);
+                if (annotation == null) continue;
+                annotations.add(annotation);
             }
         }
         return annotations;
     }
+
+    private JavaType.@Nullable Annotation annotationValue(Attribute.Compound compound) {
+        JavaType.FullyQualified annotType = TypeUtils.asFullyQualified(type(compound.type));
+        if (annotType == null) {
+            return null;
+        }
+        List<JavaType.AnnotationValue> annotationValues = new ArrayList<>();
+        for (Pair<Symbol.MethodSymbol, Attribute> attr : compound.values) {
+            JavaType.AnnotationValue annotationValue = new JavaType.AnnotationValue(
+                    methodDeclarationType(attr.fst, annotType), annotationAttributeValue(attr.snd.getValue()));
+            annotationValues.add(annotationValue);
+        }
+        return new JavaType.Annotation(annotType, annotationValues);
+    }
+
+    private @Nullable Object annotationAttributeValue(Object value) {
+        if (value instanceof Symbol.VarSymbol) {
+            return variableType((Symbol.VarSymbol) value);
+        } else if (value instanceof Type.ClassType) {
+            return type((Type.ClassType) value);
+        } else if (value instanceof Attribute.Array) {
+            List<@Nullable Object> list = new ArrayList<>();
+            for (Attribute attribute : ((Attribute.Array) value).values) {
+                list.add(annotationAttributeValue(attribute));
+            }
+            return list;
+        } else if (value instanceof Attribute.Class) {
+            return type(((Attribute.Class) value).classType);
+        } else if (value instanceof Attribute.Compound) {
+            return annotationValue((Attribute.Compound) value);
+        } else if (value instanceof Attribute.Constant) {
+            return annotationAttributeValue(((Attribute.Constant) value).value);
+        } else if (value instanceof Attribute.Enum) {
+            return annotationAttributeValue(((Attribute.Enum) value).value);
+        } else if (value instanceof List<?>) {
+            List<@Nullable Object> list = new ArrayList<>();
+            for (Object o : ((List<?>) value)) {
+                list.add(annotationAttributeValue(o));
+            }
+            return list;
+        } else if (value instanceof String) {
+            return value;
+        } else if (value instanceof Boolean) {
+            return value;
+        } else if (value instanceof Integer) {
+            return value;
+        }
+        return value;
```

Comment:
Thanks @knutwannheden! This looks much better with the actual annotation value!
Is there a reason this change is not applied to Java 8 and 11 parsers? I only need it for 17 (and possibly 21 later) but I'm curious.

Reply:
No reason other than I typically only do it for one parser first and then once I geel I am ready I update the other parsers 😇